-Pjlink-image fails with Java 14 on Helidon MP quickstart

Running `mvn clean package -Pjlink-image` with Java 14 (Java 11 is OK) on MP quickstart fails with:

```
[INFO]   checking jersey-common-2.31.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  45.105 s
[INFO] Finished at: 2020-06-12T09:35:40-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.helidon.build-tools:helidon-maven-plugin:2.0.0-RC1:jlink-image (jlink-image) on project helidon-quickstart-mp: Image generation error: Unhandled dependency: jersey-common-2.31.jar -> jdk8internals -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.helidon.build-tools:helidon-maven-plugin:2.0.0-RC1:jlink-image (jlink-image) on project helidon-quickstart-mp: Image generation error
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: Image generation error
    at io.helidon.build.maven.link.JLinkImageMojo.execute (JLinkImageMojo.java:138)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.IllegalStateException: Unhandled dependency: jersey-common-2.31.jar -> jdk8internals
    at io.helidon.linker.JavaDependencies.handleJdepsResultLine (JavaDependencies.java:156)
    at io.helidon.linker.JavaDependencies.handleJdepsResultLine (JavaDependencies.java:154)
    at io.helidon.linker.JavaDependencies.lambda$addJar$5 (JavaDependencies.java:140)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:183)
    at java.util.stream.ReferencePipeline$2$1.accept (ReferencePipeline.java:177)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:195)
    at java.util.Spliterators$ArraySpliterator.forEachRemaining (Spliterators.java:948)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:484)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:474)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:150)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:173)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:497)
    at io.helidon.linker.JavaDependencies.addJar (JavaDependencies.java:140)
    at io.helidon.linker.JavaDependencies.lambda$collect$1 (JavaDependencies.java:87)
    at java.util.ArrayList$ArrayListSpliterator.forEachRemaining (ArrayList.java:1624)
    at java.util.stream.Streams$ConcatSpliterator.forEachRemaining (Streams.java:735)
    at java.util.stream.ReferencePipeline$Head.forEach (ReferencePipeline.java:658)
    at io.helidon.linker.JavaDependencies.collect (JavaDependencies.java:83)
    at io.helidon.linker.JavaDependencies.collect (JavaDependencies.java:72)
    at io.helidon.linker.Application.javaDependencies (Application.java:78)
    at io.helidon.linker.Linker.collectJavaDependencies (Linker.java:158)
    at io.helidon.linker.Linker.link (Linker.java:118)
    at io.helidon.build.maven.link.JLinkImageMojo.execute (JLinkImageMojo.java:136)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:564)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

To reproduce:

```
echo $JAVA_HOME
/Library/Java/JavaVirtualMachines/jdk-14.0.1.jdk/Contents/Home

mvn archetype:generate -DinteractiveMode=false     -DarchetypeGroupId=io.helidon.archetypes     -DarchetypeArtifactId=helidon-quickstart-mp -DarchetypeVersion=2.0.0-RC1 -DgroupId=io.helidon.examples     -DartifactId=helidon-quickstart-mp     -Dpackage=io.helidon.examples.quickstart.mp

mvn -e clean package -Pjlink-image
```

--- a/helidon-linker/src/main/java/io/helidon/linker/Application.java
+++ b/helidon-linker/src/main/java/io/helidon/linker/Application.java
@@ -23,7 +23,6 @@ import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
-import java.util.Locale;
 import java.util.Set;
 import java.util.stream.Stream;
 
@@ -32,6 +31,7 @@ import io.helidon.build.util.Log;
 import io.helidon.linker.util.JavaRuntime;
 
 import static io.helidon.linker.util.Constants.DIR_SEP;
+import static java.util.jar.Attributes.Name.IMPLEMENTATION_VERSION;
 
 /**
  * A Helidon application supporting Java dependency collection and installation into a Java Home.
@@ -224,25 +224,12 @@ public final class Application implements ResourceContainer {
     }
 
     private String extractHelidonVersion() {
-        return versionFromFileName(classPath.stream()
-                                            .map(Jar::name)
-                                            .filter(name -> name.startsWith(HELIDON_JAR_NAME_PREFIX))
-                                            .findFirst()
-                                            .orElseThrow(() -> new IllegalStateException("No helidon jars found!")));
-    }
-
-
-    static String versionFromFileName(String fileName) {
-        final int lastDot = fileName.lastIndexOf('.');
-        if (lastDot >= 0) {
-            final String name = fileName.substring(0, lastDot);
-            final boolean snapshot = name.toUpperCase(Locale.ENGLISH).endsWith(SNAPSHOT);
-            final String nonSnapshotName = snapshot ? name.substring(0, name.length() - SNAPSHOT.length()) : name;
-            final int lastDash = nonSnapshotName.lastIndexOf('-');
-            if (lastDash >= 0 && Character.isDigit(name.charAt(lastDash + 1))) {
-                return name.substring(lastDash + 1);
-            }
-        }
-        return UNKNOWN_VERSION;
+        return classPath.stream()
+                        .filter(jar -> jar.name().startsWith(HELIDON_JAR_NAME_PREFIX))
+                        .filter(jar -> jar.manifest() != null)
+                        .filter(jar -> jar.manifest().getMainAttributes().containsKey(IMPLEMENTATION_VERSION))
+                        .map(jar -> jar.manifest().getMainAttributes().getValue(IMPLEMENTATION_VERSION))
+                        .findFirst()
+                        .orElse(UNKNOWN_VERSION);
     }
 }

--- a/helidon-linker/src/main/java/io/helidon/linker/Jar.java
+++ b/helidon-linker/src/main/java/io/helidon/linker/Jar.java
@@ -220,6 +220,15 @@ public final class Jar implements ResourceContainer {
     }
 
     /**
+     * Returns the manifest, if present.
+     *
+     * @return The manifest or {@code null} if not present.
+     */
+    public Manifest manifest() {
+        return manifest;
+    }
+
+    /**
      * Returns the entries in this jar.
      *
      * @return The entries.
@@ -328,10 +337,10 @@ public final class Jar implements ResourceContainer {
         if (Constants.OS.isPosix()) {
             try {
                 Files.setPosixFilePermissions(targetFile, Set.of(
-                    PosixFilePermission.OWNER_READ,
-                    PosixFilePermission.OWNER_WRITE,
-                    PosixFilePermission.GROUP_READ,
-                    PosixFilePermission.OTHERS_READ
+                        PosixFilePermission.OWNER_READ,
+                        PosixFilePermission.OWNER_WRITE,
+                        PosixFilePermission.GROUP_READ,
+                        PosixFilePermission.OTHERS_READ
                 ));
             } catch (IOException e) {
                 Log.warn("Unable to set %s read-only: %s", e.getMessage());

--- a/helidon-linker/src/main/java/io/helidon/linker/JavaDependencies.java
+++ b/helidon-linker/src/main/java/io/helidon/linker/JavaDependencies.java
@@ -54,6 +54,7 @@ public final class JavaDependencies {
         "split package", JavaDependencies::split,
         "not found", JavaDependencies::ignore,
         "unnamed module", JavaDependencies::ignore,
+        "jdk8internals", JavaDependencies::debug,
         "JDK removed internal API", JavaDependencies::debug
     );
 

--- a/helidon-linker/src/test/java/io/helidon/linker/ApplicationTest.java
+++ b/helidon-linker/src/test/java/io/helidon/linker/ApplicationTest.java
@@ -16,11 +16,16 @@
 
 package io.helidon.linker;
 
+import java.nio.file.Path;
+
+import io.helidon.build.test.TestFiles;
+
 import org.junit.jupiter.api.Test;
 
-import static io.helidon.linker.Application.versionFromFileName;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.notNullValue;
 
 /**
  * Unit test for class {@link Application}.
@@ -28,12 +33,12 @@ import static org.hamcrest.Matchers.is;
 class ApplicationTest {
 
     @Test
-    void testVersionParsing() {
-        assertThat(versionFromFileName("helidon-config-1.4.2.jar"), is("1.4.2"));
-        assertThat(versionFromFileName("helidon-config-1.4.3-SNAPSHOT.jar"), is("1.4.3-SNAPSHOT"));
-        assertThat(versionFromFileName("helidon-config-2.0.jar"), is("2.0"));
-        assertThat(versionFromFileName("helidon-config-2.0-SNAPSHOT.jar"), is("2.0-SNAPSHOT"));
-        assertThat(versionFromFileName("helidon-config.jar"), is("0.0.0"));
+    void testHelidonVersion() {
+        Path mainJar = TestFiles.helidonMpJar();
+        Application app = Application.create(mainJar);
+        String version = app.helidonVersion();
+        assertThat(version, is(notNullValue()));
+        assertThat(version, is(not("0.0.0")));
     }
 
     @Test

